File: mingle-io/ruby/lib/mingle/io/stream.rb
require 'bitgirder/core'
require 'bitgirder/io'
require 'mingle/io'

module Mingle
module Io
module Stream

MESSAGE_VERSION1 = 1

TYPE_CODE_HEADERS = 1
TYPE_CODE_MESSAGE_BODY = 2

class Message < BitGirder::Core::BitGirderClass

    bg_attr :headers, 
            :processor => lambda { |val|
                Mingle::Io::Headers.as_headers( val )
            },
            :default => lambda { Mingle::Io::Headers.as_headers( {} ) }

    bg_attr :body, :default => "".freeze
end

class Connection < BitGirder::Core::BitGirderClass

    include Mingle

    bg_attr :reader
    bg_attr :writer

    private
    def impl_initialize

        @enc = Io::Encoder.new( @writer )
        @dec = Io::Decoder.new( @reader )
    end

    private
    def write_headers( hdrs )

        @enc.write_int32( TYPE_CODE_HEADERS )
        @enc.write_headers( hdrs )
    end

    private
    def write_body( body )
        
        @enc.write_int32( TYPE_CODE_MESSAGE_BODY )
        @enc.write_int64( sz = body.bytesize )
        @writer.write( body ) unless sz == 0
    end

    public
    def write_message( msg )

        msg = Message.as_instance( msg )

        @enc.write_int32( MESSAGE_VERSION1 )
        write_headers( msg.headers )
        write_body( msg.body )
        @writer.flush
    end

    private
    def read_body
        sz = @dec.read_int64
        @dec.read_full( sz )
    end

    public
    def read_message
        
        @dec.expect_version( MESSAGE_VERSION1, "message" )

        @dec.expect_type_code( TYPE_CODE_HEADERS )
        hdrs = @dec.read_headers

        @dec.expect_type_code( TYPE_CODE_MESSAGE_BODY )
        body = read_body
        
        Message.new( :headers => hdrs, :body => body )
    end

end

class MinglePeer < BitGirder::Core::BitGirderClass

    include Mingle
    
    private_class_method :new

    bg_attr :proc_builder

    private
    def start
        @peer = @proc_builder.popen( "r+" )
        @conn = Connection.new( :reader => @peer, :writer => @peer )
    end

    public
    def exchange_message( msg )
        
        @conn.write_message( msg )
        @conn.read_message
    end

    public
    def await_exit( opts = { :expect_success => true } )

        not_nil( opts, :opts )

        msg = nil

        begin
            pid, stat = 
                BitGirder::Io.debug_wait2( :pid => @peer.pid, :name => :peer )
    
            unless stat.success?
                msg = "Peer #{pid} exited with non-success #{stat.exitstatus}"
            end
        rescue Errno::ECHILD
            msg = "Peer #{pid} appears to have exited already"
        end
        
        if msg
            if opts[ :expect_success ] then raise msg else warn msg end
        end
    end

    def self.open( *argv )
        
        res = self.send( :new, *argv )
        res.send( :start )

        res
    end
end

end
end
end

File: mingle-io/go/lib/mingle/io/stream/stream.go
package stream

import (
    "io"
//    "log"
    mgio "mingle/io"
)

const (
    messageVersion1 = int32( 1 )
    typeCodeHeaders = int32( 1 )
    typeCodeMessageBody = int32( 2 )
)

type BodyWriter func( w io.Writer ) error

type MessageReader func( msg *mgio.Headers, sz int64, r io.Reader ) error

func writeMessage(
    msg *mgio.Headers, sz int64, bw BodyWriter, w io.Writer ) error {
    if err := mgio.WriteInt32( messageVersion1, w ); err != nil { return err }
    if err := mgio.WriteInt32( typeCodeHeaders, w ); err != nil { return err }
    if err := mgio.WriteHeaders( msg, w ); err != nil { return err }
    if err := mgio.WriteInt32( typeCodeMessageBody, w ); err != nil { 
        return err 
    }
    if err := mgio.WriteInt64( sz, w ); err != nil { return err }
    if sz > 0 { return bw( w ) }
    return nil
}

func readMessage( rd io.Reader, mr MessageReader ) error {
    if err := mgio.ReadVersion( messageVersion1, "message", rd ); err != nil { 
        return err 
    }
    if err := mgio.ExpectTypeCode( typeCodeHeaders, rd ); err != nil { 
        return err
    }
    hdrs, err := mgio.ReadHeaders( rd )
    if err != nil { return err }
    if err = mgio.ExpectTypeCode( typeCodeMessageBody, rd ); err != nil {
        return err
    }
    var sz int64
    if sz, err = mgio.ReadInt64( rd ); err != nil { return err }
    rd = io.LimitReader( rd, sz )
    return mr( hdrs, sz, rd )
}

type Connection interface {
    WriteMessage( msg *mgio.Headers, sz int64, w BodyWriter ) error
    ReadMessage( r MessageReader ) error
    Close() error
}

type connectionImpl struct {
    rd io.ReadCloser
    wr io.WriteCloser
}

func ( c *connectionImpl ) Close() error {
    err1 := c.rd.Close()
    err2 := c.wr.Close()
    if err1 == nil { return err2 }
    return err1
}

func ( c *connectionImpl ) WriteMessage(
    msg *mgio.Headers, sz int64, w BodyWriter ) error {
    return writeMessage( msg, sz, w, c.wr )
}

func ( c *connectionImpl ) ReadMessage( r MessageReader ) error {
    return readMessage( c.rd, r )
}

func NewConnection( rd io.ReadCloser, wr io.WriteCloser ) Connection {
    return &connectionImpl{ rd, wr }
}

func OpenPipe() ( cli, srv Connection ) {
    p1Rd, p1Wr := io.Pipe()
    p2Rd, p2Wr := io.Pipe()
    return NewConnection( p1Rd, p2Wr ), NewConnection( p2Rd, p1Wr )
}

func WriteMessageBytes( 
    conn Connection, msg *mgio.Headers, body []byte ) error {
    sender := func( w io.Writer ) error {
        _, err := w.Write( body )
        return err
    }
    return conn.WriteMessage( msg, int64( len( body ) ), sender )
}

func WriteMessageNoBody( conn Connection, msg *mgio.Headers ) error {
    return WriteMessageBytes( conn, msg, []byte{} )
}

func ReadMessageBytes(
    conn Connection ) ( msg *mgio.Headers, body []byte, err error ) {
    reader := func( msgIn *mgio.Headers, sz int64, r io.Reader ) error {
        msg = msgIn
        body = make( []byte, sz )
        _, err = io.ReadFull( r, body )
        return err
    }
    err = conn.ReadMessage( reader )
    return
}
